### ISSUE TITLE
Tile fails:  Couldn't open file anaconda powershell

### ISSUE THREAD

**Issue by lspaete** [PDAL/PDAL#3344]
when running tile from ananconda powershell prompt  get the error "PDAL: writers.las: Couldn't open file 'D:\PineCo\tiled\PineCo_498_5101.laz' for output."

pdal --version
pdal 2.2.0 (git-version: Release)

build
Microsoft Windows Server 2019 Standard  Multiprocessor Free 64-bit

**Comment by hobu** [comment]
My guess is this is some kind of limit on number of files open at once. Maybe `pdal tile` is leaving handles open.

**Comment by abellgithub** [comment]
`pdal tile` leaves open all files as it writes as there is no support in PDAL for updating files once closed. You have to make sure that you have enough file handles available to support the number of output files.

**Comment by hobu** [comment]
> You have to make sure that you have enough file handles available to support the number of output files.

hmm... There's no ulimit-like thing in Windows. It's per-system, right?

**Comment by abellgithub** [comment]
A quick Google suggests that the limit is huge, so maybe this isn't the problem? Would need a command and more information to diagnose.

**Comment by lspaete** [comment]
>pdal tile  --input "D:\PineCo\tile\*.LAZ" --output "D:\PineCo\tiled\PineCo_#.laz" --origin_x 0 --origin_y 0 --buffer 20

I have tried it on several datasets and the error always pops up around tile #500. This makes me think it is a handle issue. Perhaps coming from python?

**Comment by abellgithub** [comment]
This says 512:

https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/setmaxstdio

**Comment by lspaete** [comment]
The way i account for the difference is  other handles not associated with that particular process are already in use. Would that make sense?

**Comment by hobu** [comment]
> The way i account for the difference is other handles not associated with that particular process are already in use. Would that make sense?

Yes.

**Comment by hobu** [comment]
> By default, up to 512 files can be open simultaneously at the stream I/O level. This level includes files opened and accessed using the fopen, fgetc, and fputc family of functions. The limit of 512 open files at the stream I/O level can be increased to a maximum of 8,192 by use of the _setmaxstdio function.

Even if we bump the limit, if there are more than 8k files in the directory it is going to fail again.

**Comment by abellgithub** [comment]
Well, 8k output files is many more than 512.

**Comment by lspaete** [comment]
is this something i can implement in anaconda or does it need to be built into pdal?

**Comment by abellgithub** [comment]
It needs to be done in PDAL, but you can get around this a bit by doing sub-processing, if you want. For example, first tile into 16 files, then tile each of those 16 into 500 subtiles, for example.

**Comment by newcomb-d** [comment]
Would this be the reason for the pdal_wrench failure in https://github.com/qgis/QGIS/issues/53970  ?

**Comment by hobu** [comment]
yes, most likely

**Comment by wonder-sk** [comment]
> Would this be the reason for the pdal_wrench failure in https://github.com/qgis/QGIS/issues/53970 ?

Not really, because pdal_wrench does not use the implementation from `pdal tile`. The implementation in pdal_wrench is based on untwine, and it should not be prone to exhaustion of file handles in operating system.

**Comment by newcomb-d** [comment]
Just an FYI, I also see a similar limitation in the number of files for r.in.pdal with GRASS 8.2.1 on Linux
________________________________
From: Martin Dobias ***@***.***>
Sent: Wednesday, August 9, 2023 3:02 PM
To: PDAL/PDAL ***@***.***>
Cc: Newcomb, Doug ***@***.***>; Comment ***@***.***>
Subject: [EXTERNAL] Re: [PDAL/PDAL] pdal tile exhausts open file handles on Windows (#3344)




 This email has been received from outside of DOI - Use caution before clicking on links, opening attachments, or responding.



Would this be the reason for the pdal_wrench failure in qgis/QGIS#53970<https://github.com/qgis/QGIS/issues/53970> ?

Not really, because pdal_wrench does not use the implementation from pdal tile. The implementation in pdal_wrench is based on untwine, and it should not be prone to exhaustion of file handles in operating system.
